src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +211,560 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
+								 uint32 taskIdIndex)
+{
+	Query *copiedQuery = copyObject(originalQuery);
+	RangeTblEntry *copiedInsertRte = ExtractInsertRangeTableEntry(copiedQuery);
+	RangeTblEntry *copiedSubqueryRte = ExtractSelectRangeTableEntry(copiedQuery);
+	Query *copiedSubquery = (Query *) copiedSubqueryRte->subquery;
+
+	uint64 shardId = shardInterval->shardId;
+	Oid distributedTableId = shardInterval->relationId;
+
+	RelationRestrictionContext *copiedRestrictionContext =
+		CopyRelationRestrictionContext(restrictionContext);
+
+	StringInfo queryString = makeStringInfo();
+	ListCell *restrictionCell = NULL;
+	Task *modifyTask = NULL;
+	List *selectPlacementList = NIL;
+	uint64 selectAnchorShardId = INVALID_SHARD_ID;
+	uint64 jobId = INVALID_JOB_ID;
+	List *insertShardPlacementList = NULL;
+	List *intersectedPlacementList = NULL;
+	bool routerPlannable = false;
+	bool upsertQuery = false;
+
+	/* grab shared metadata lock to stop concurrent placement additions */
+	LockShardDistributionMetadata(shardId, ShareLock);
+
+	/*
+	 * Replace the partitioning qual parameter value in all baserestrictinfos.
+	 * Note that this has to be done on a copy, as the walker modifies in place.
+	 */
+	foreach(restrictionCell, copiedRestrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *restriction = lfirst(restrictionCell);
+		List *originalBaserestrictInfo = restriction->relOptInfo->baserestrictinfo;
+
+		originalBaserestrictInfo =
+			(List *) InstantiatePartitionQual((Node *) originalBaserestrictInfo,
+											  shardInterval);
+	}
+
+	/*
+	 * Use router select planner to decide on whether we can push down the query
+	 * or not. If we can, we also rely on the side-effects that all RTEs have been
+	 * updated to point to the relevant nodes and selectPlacementList is determined.
+	 */
+	routerPlannable = RouterSelectQuery(copiedSubquery, copiedRestrictionContext,
+										&selectPlacementList, &selectAnchorShardId);
+
+	if (!routerPlannable)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot perform distributed planning for the given "
+							   "modification"),
+						errdetail("Select query cannot be pushed down to the worker.")));
+	}
+
+	/* Ensure that we have INSERTed table's placement exists on the same worker */
+	insertShardPlacementList = FinalizedShardPlacementList(shardId);
+	intersectedPlacementList = IntersectPlacementList(insertShardPlacementList,
+													  selectPlacementList);
+
+	if (list_length(insertShardPlacementList) != list_length(intersectedPlacementList))
+	{
+		ereport(DEBUG2, (errmsg("could not generate task for target shardId: %ld",
+								shardId),
+						 errdetail("Insert query hits %d placements, Select query "
+								   "hits %d placements and only %d of those placements match.",
+								   list_length(insertShardPlacementList),
+								   list_length(selectPlacementList),
+								   list_length(intersectedPlacementList))));
+
+		return NULL;
+	}
```

Comment:
Yes, let me also share the output which you could see that the logic allows to push down:

```SQL

INSERT INTO raw_events_second(user_id) 
SELECT user_id
FROM   raw_events_first 
WHERE  user_id IN (SELECT user_id 
                   FROM   raw_events_second 
                   WHERE  user_id = 2);

DEBUG:  distributed statement: INSERT INTO public.raw_events_second_102024 AS citus_table_alias (user_id) SELECT user_id FROM public.raw_events_first_102023 raw_events_first WHERE (user_id IN (SELECT raw_events_second.user_id FROM public.raw_events_second_102027 raw_events_second WHERE (raw_events_second.user_id = 2)))

DEBUG:  distributed statement: INSERT INTO public.raw_events_second_102025 AS citus_table_alias (user_id) SELECT user_id FROM public.raw_events_first_102023 raw_events_first WHERE (user_id IN (SELECT raw_events_second.user_id FROM public.raw_events_second_102027 raw_events_second WHERE (raw_events_second.user_id = 2)))

DEBUG:  distributed statement: INSERT INTO public.raw_events_second_102026 AS citus_table_alias (user_id) SELECT user_id FROM public.raw_events_first_102023 raw_events_first WHERE (user_id IN (SELECT raw_events_second.user_id FROM public.raw_events_second_102027 raw_events_second WHERE (raw_events_second.user_id = 2)))

DEBUG:  distributed statement: INSERT INTO public.raw_events_second_102027 AS citus_table_alias (user_id) SELECT user_id FROM public.raw_events_first_102023 raw_events_first WHERE (user_id IN (SELECT raw_events_second.user_id FROM public.raw_events_second_102027 raw_events_second WHERE (raw_events_second.user_id = 2))) 
```

Note that we have 4 shards - two replicas on two nodes.

In simple words, per target shard interval, we could always find SELECT query placements that are on the same worker node. So, we allow that task to be pushed down.

However, with that we unnecessaryly insert data to the unrelated shards. We probably need one another check to prevent that?